### ISSUE TITLE
MacOS - Terraform v0.14 - docker_container volume mount - host_path absolute path malfunction -> Shared Volume Issue

### ISSUE THREAD

**Issue by EpicApex** [kreuzwerker/terraform-provider-docker#179]
### Terraform (and docker Provider) Version

Terraform v0.14.0
on darwin_amd64
+ provider registry.terraform.io/cyrilgdn/rabbitmq v1.5.1
+ provider registry.terraform.io/hashicorp/null v3.1.0
+ provider registry.terraform.io/kreuzwerker/docker v2.11.0

#### main.tf
```hcl
required_providers {
  docker = {
    source = "kreuzwerker/docker"
    version = "2.11.0"
  }
  rabbitmq = {
    source = "cyrilgdn/rabbitmq"
    version = "1.5.1"
  }
 }
}

provider "docker" {
  	host = "unix:///var/run/docker.sock"
}

resource "docker_container" "nginx" {
  image = docker_image.nginx.name
  name = "nginx-local-server"
  
  volumes {
    host_path = docker_volume.shared_volume.name
    container_path = var.nginx_container_volume_mount
  }

  ports {
    internal = var.nginx_internal_port
    external = var.nginx_external_port
    ip = var.nginx_ip
  }
}

resource "docker_volume" "shared_volume" {
    name = var.nginx_host_volume
}
```

#### variables.tf
```
variable "nginx_host_volume" {
    description = "Nginx docker host volume"
    type = string
    default = "/Users/omri/dev/Local_Infrastracture/configs/nginx_conf.d"
  } 

```
#### File Structure

main.tf
variables.tf ( contains host path absolute path )

### Expected Behavior

Infra should be built without any failures, by digesting the host_path with absolute path - no matter where it is coming from - variables.tf or main.tf.

### Actual Behaviour

Error: Unable to create volume: Error response from daemon: create /Users/omri/dev/Local_Infrastracture/configs/nginx_conf.d: "/Users/omri/dev/Local_Infrastracture/configs/nginx_conf.d" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path

### Steps to Reproduce

1. terraform init
2. terraform apply

### Important Factoids

**Didnt work**
1. Injected absolute path manually without variables. 
2. Injected relative path manually without variables. 
3. Injected absolute path via variables
4. Injected relative path via variables 

## After several more tryouts I have located the issue to be related to "Docker shared Volume"
resource "docker_volume" "shared_volume"
{
    name = var.nginx_host_volume
}

- Once the above is commented out - absolute path is working in conjunction with variables.tf parameters.

### References

https://github.com/kreuzwerker/terraform-provider-docker/issues/55

**Comment by suzuki-shunsuke** [comment]
Please use the issue template.
https://github.com/kreuzwerker/terraform-provider-docker/issues/new?assignees=&labels=&template=Bug_Report.md

**Comment by suzuki-shunsuke** [comment]
```
Error: Unable to create volume: Error response from daemon: create /Users/omri/dev/terraform-dev/Local_Infrastracture/configs/nginx_conf.d: "/Users/omri/dev/terraform-dev/Local_Infrastracture/configs/nginx_conf.d" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path
```

```tf
resource "docker_volume" "shared_volume"
{
    name = var.nginx_host_volume
}

variable "nginx_host_volume"
  {
    description = "Nginx docker host volume"
    type = string
    default = "/Users/omri/dev/Local_Infrastracture/configs/nginx_conf.d"
  } 
```

In your code, the volume path is `/Users/omri/dev/Local_Infrastracture/configs/nginx_conf.d`, but it's wrong.

**Comment by EpicApex** [comment]
> ```
> Error: Unable to create volume: Error response from daemon: create /Users/omri/dev/terraform-dev/Local_Infrastracture/configs/nginx_conf.d: "/Users/omri/dev/terraform-dev/Local_Infrastracture/configs/nginx_conf.d" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path
> ```
> 
> ```terraform
> resource "docker_volume" "shared_volume"
> {
>     name = var.nginx_host_volume
> }
> 
> variable "nginx_host_volume"
>   {
>     description = "Nginx docker host volume"
>     type = string
>     default = "/Users/omri/dev/Local_Infrastracture/configs/nginx_conf.d"
>   } 
> ```
> 
> In your code, the volume path is `/Users/omri/dev/Local_Infrastracture/configs/nginx_conf.d`, but it's wrong.
> 
> ref. https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/volume

Hi @suzuki-shunsuke 

I fail to understand **why** is it wrong, please specify how you think this should be written - the provided docs are known to me and have no direction as to the **why**. 

(I've edited the path to - /Users/omri/dev/Local_Infrastracture/configs/nginx_conf.d --> And the same error appears)

Also if you can share a working docker shared volume example / repo that would be amazing because this is my suspicion here.

**Comment by suzuki-shunsuke** [comment]
```console
$ terraform version 
Terraform v0.14.10
+ provider registry.terraform.io/kreuzwerker/docker v2.11.0
```

docker_volume.name is a Docker Volume name, which isn't a path.

```tf
resource "docker_volume" "test" {
    name = "/test"
}
```

is equivalent to

```console
$ docker volume create /test
Error response from daemon: create /test: "/test" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path
```

I don't understand what you want well, but I show two examples.

## Example 1

Share a volume within two containers.

```tf
terraform {
  required_providers {
    docker = {
      source  = "kreuzwerker/docker"
      version = "2.11.0"
    }
  }
}

provider "docker" {
}

resource "docker_container" "alpine" {
  name  = "alpine"
  image = docker_image.alpine.latest
  volumes {
    volume_name    = docker_volume.test.name
    container_path = "/test-volume"
  }
  command = ["tail", "-f", "/dev/null"]
}

resource "docker_container" "alpine2" {
  name  = "alpine2"
  image = docker_image.alpine.latest
  volumes {
    volume_name    = docker_volume.test.name
    container_path = "/test-volume"
  }
  command = ["tail", "-f", "/dev/null"]
}

resource "docker_image" "alpine" {
  name = "alpine:3.13.0"
}

resource "docker_volume" "test" {
  name = "test"
}
```

## Example 2

Specify the absolute path to `volumes.host_path`.

```tf
terraform {
  required_providers {
    docker = {
      source  = "kreuzwerker/docker"
      version = "2.11.0"
    }
  }
}

provider "docker" {
}

resource "docker_container" "alpine" {
  name  = "alpine"
  image = docker_image.alpine.latest
  volumes {
    host_path      = "/Users/foo/Documents/test/terraform/terraform-provider-docker/issue-179/test-volume"
    container_path = "/test-volume"
  }
  command = ["tail", "-f", "/dev/null"]
}

resource "docker_image" "alpine" {
  name = "alpine:3.13.0"
}
```

**Comment by EpicApex** [comment]
Firstly thank you for the examples.

But as I've showed in my examples 'host path' as defined [here](https://registry.terraform.io/providers/kreuzwerker/docker/latest/docs/resources/container) which resulted with the error in the first place.

If I'm not using 'host-path' correctly as stated in the docs, could you provide an example of sort?

Cheers,

**Comment by alexkunin** [comment]
@EpicApex, let me intervene, I think I understand what's wrong and how to fix it.

Your code looks like this (removed parts, irrelevant to this issue):

```hcl
resource "docker_volume" "shared_volume"
{
    name = var.nginx_host_volume
}

resource "docker_container" "nginx"
{
  volumes {
    host_path = docker_volume.shared_volume.name
    container_path = var.nginx_container_volume_mount
  }
}
```
But this is how to express in terms of resources what you actually mean:
```hcl
resource "docker_container" "nginx"
{
  volumes {
    host_path = var.nginx_host_volume
    container_path = var.nginx_container_volume_mount
  }
}
```
I think the problem comes from a bit overloaded term "volume":

1. Docker has volume object -- one that you can see after doing `docker volume ls`
2. also, when container is created, Docker mounts host file/directory **or** volume object by using `--volume` option (usually it's short form`-v` is used)
3. finally, there is `kreuzwerker/terraform-provider-docker` resource `docker_volume` which is exactly the same concept as p.1 above, and also can be used as one of possible inputs for p.2.

But your case in p.2 is not volume object, it's actually a host file/directory case, and thus it does not require separate creation of volume object.

BTW, in `docker_service` resource all these things are called mounts, i.e. you mount a `docker_volume` resource or host file/directory, and this terminology is much more clear. But `kreuzwerker/terraform-provider-docker` is not to blame for this, as far as I can tell guys just follow Docker API terminology.

Basically, this is the same what @suzuki-shunsuke said, I'm just trying to explain it differently.

**Comment by Junkern** [comment]
I am closing this, as it is not an issue.
